test/lit/basic/extra-branch-values.wast

```diff
@@ -0,0 +1,211 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+
+;; RUN: wasm-opt %s -all -S -o - | filecheck %s
+
+(module
+  ;; CHECK:      (func $br_on_null-one (type $3) (result i32)
+  ;; CHECK-NEXT:  (local $scratch nullref)
+  ;; CHECK-NEXT:  (local $scratch_1 i32)
+  ;; CHECK-NEXT:  (local $scratch_2 i32)
+  ;; CHECK-NEXT:  (local $scratch_3 (ref none))
+  ;; CHECK-NEXT:  (block $block (result i32)
+  ;; CHECK-NEXT:   (block $block0
+  ;; CHECK-NEXT:    (local.set $scratch_1
+  ;; CHECK-NEXT:     (block (result i32)
+  ;; CHECK-NEXT:      (local.set $scratch_2
+  ;; CHECK-NEXT:       (i32.const 0)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (local.set $scratch
+  ;; CHECK-NEXT:       (ref.null none)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (local.get $scratch_2)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (local.set $scratch_3
+  ;; CHECK-NEXT:     (br_on_null $block0
+  ;; CHECK-NEXT:      (local.get $scratch)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (local.get $scratch_1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (drop
+  ;; CHECK-NEXT:     (local.get $scratch_3)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (br $block
+  ;; CHECK-NEXT:     (unreachable)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (local.get $scratch_1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $br_on_null-one (result i32)
+    block (result i32)
+      i32.const 0
+      ref.null none
+      br_on_null 0
+      unreachable
+    end
+  )
```

Comment:
This does look like quite a lot. Maybe lowering the `br_on_null` itself makes sense? That is
```wat
(block (result i32)
  (local.set $scratch_value (i32.const 0))
  (local.set $scratch_ref (ref.null none))
  (if
    (ref.is_null (local.get $scratch_ref))
    (br 0 (local.get $scratch_value))
    (tuple.make 2
      (local.get $scratch_value)
      (local.get $scratch_ref)
    )
  )
)
```
Hopefully TupleOptimization can then remove the tuple in many cases.

Reply:
That's a good idea. I'll play around with it.

Reply:
Actually, I'm not sure this would help much. A big simplification in your example comes from ordering the `local.set $scratch_value` before the evaluation of the value under test, which is not possible in general. Another simplification is that you don't use a separate scratch local to capture the output when the branch is not taken, but that means you would have to insert a `ref.as_non_null` when reassembling the output tuple. For br_on_non_null and br_on_cast, doing a similar lowering would require inserting a cast in the trampoline instead, which means the trampolines would have to be keyed on the arity and cast type, not just the arity.

I think the current code is as simple as we can get without building a bunch of optimizations into IRBuilder to exploit special cases.